File: python/servo.py
'''
 Program: servo.py
 Description: Generates a PWM signal to control a servo motor
 Author: Aldo Nunez
'''
import sys
import mraa as m

'''
 Program: turnServo
 Parameters: p	- Reference to pwm
             t  - Period in micro seconds
             pw - Pulse width
 Output:     None
 Description: Enable, set the period and send
            the command to generate pwm signal 				
'''
def turn_servo ( p, t, d ):
    p.period_us ( t )
    p.enable ( True ) 
    p.pulsewidth ( d )
    return;

def main ( argv ):
    try:
        n =  sys.argv [ 1 ]
    except:
        print ( "<Usage>: " + sys.argv [ 0 ] + " <DIR>" )
        print ( "<DIR>: CW or CCW")
        sys.exit()
    
    PORT = 3
    T = 20000           # Period: 20 msec; f = 50 Hz
    CLKWISE = 1.0e-3
    CCLKWISE = 2.0e-3


    pwm = m.Pwm ( PORT )
    pwm.period_us ( T )
    turn_dir = sys.argv [ 1 ]

    if turn_dir == 'CW':
        turn_servo ( pwm, T, CLKWISE )
    elif turn_dir == 'CCW':
        turn_servo ( pwm, T, CCLKWISE )
    else:
        print ( "<Usage>: " + sys.argv [ 0 ] + " <DIR>" )
        print ( "<DIR>: CW or CCW")
        sys.exit ()

    print ( "MRAA Version: " + m.getVersion () )
    print ( "Platform: " + m.getPlatformName () )
    print ( "Port Number: " +  str ( PORT) )
    print ( "Clockwise" if turn_dir == "CW"  else "Counter clockwise" )
    c = raw_input ( "" )

if __name__ == "__main__":
    main ( sys.argv [ 1: ] )

File: c/src/servo.c
/*
** Name: servo.c
** Description: Generate a PWM signal to control a servo motor.
** 				The user enters the turn direction: CW or CCW
** Author: Aldo Nunez
*/

#include <stdio.h>
#include <string.h>
#include <mraa.h>

/*
** Program: turnServo
** Parameters:	ptr	- Pointer to Pwm
**				t 	- Period in micro seconds
** 				pw 	- Pulse width
** Output:		None
** Description: Enable, set the period and send
**				the command to generate pwm signal 				
*/
void turn_servo ( mraa_pwm_context*, float, float );

enum PWM { PWM0 = 3, PWM1 = 5, PWM2 = 6, PWM3 = 9 } pwm_port; /* PWM  ports */
const int T = 20000 ; 				/* T period in usec. 20000 ~ 20 ms;  f ~ 50 hz */
const float CLKWISE = 1.0e-3;  		/* Clockwise, pulse width  1 ms */
const float CCLKWISE = 2.0e-3;		/* Counter Clockwise, pulse width 2 ms */

int
main ( int argc, char* argv [] )
{
	if ( argc < 2 )
	{
		printf ( "<Usage>: %s <DIR> \n", argv [ 0 ] );
		printf ( "DIR: <CW> or <CCW> \n" );

		return 1;
	}
	pwm_port = PWM0;
	mraa_init ();
	mraa_pwm_context pwm;
	pwm = mraa_pwm_init ( pwm_port );

	if ( pwm == NULL )
	{
        return 1;
	}

	char* dir =  argv [ 1 ];
	if ( !strcmp ( dir, "CW" ) )
	{
		turn_servo ( &pwm, T, CLKWISE );
		dir = "Clockwise";
	}
	else if ( !strcmp ( dir, "CCW" ) )
	{
		turn_servo ( &pwm, T, CCLKWISE );
		dir = "Counter clockwise";
	}
	else
	{
		printf ( "Usage: ./servo <DIR> \n" );
		printf ( "DIR: <CW> or <CCW> \n" );

		return 1;
	}

	printf ( "MRAA Version: %s \n", mraa_get_version () );
	printf ( "Platform: %s \n", mraa_get_platform_name () );
	printf ( "PWM port: %d \n", pwm_port );
	printf ( "%s \n", dir ); 

	getchar ();
	mraa_pwm_close ( pwm );

	return ( MRAA_SUCCESS );
}

void
turn_servo ( mraa_pwm_context* ptr, float t, float pw )
{
		mraa_pwm_enable ( *ptr, 1 );
		mraa_pwm_period_us (  *ptr, t ); 
		mraa_pwm_pulsewidth ( *ptr, pw );

		return;
}
